### PR TITLE
Add --no-html flag to Phoenix installer

### PR DESCRIPTION
**Pull request by stevedomin**
This allows to skip generating HTML views and including related
dependencies (phoenix_html, phoenix_live_reload).

Fixes https://github.com/phoenixframework/phoenix/issues/1192

### CODE REVIEW

File: installer/lib/phoenix_new.ex [hunk 1]
```diff
@@ -153,7 +164,8 @@ defmodule Mix.Tasks.Phoenix.New do
   def run(app, mod, path, opts) do
     db = Keyword.get(opts, :database, "postgres")
     ecto = Keyword.get(opts, :ecto, true)
-    brunch = Keyword.get(opts, :brunch, true)
+    html = Keyword.get(opts, :html, true)
+    brunch = Keyword.get(opts, :brunch, html && true)
```
Comment (josevalim) [inline installer/lib/phoenix_new.ex]: I am not sure if we should do this. I can have my javascript app all written with brunch and just use json. Maybe we do need an `--api` flag that is a shortcut for `--no-html` and `--no-brunch`?

File: installer/lib/phoenix_new.ex [hunk 2]
```diff
@@ -184,6 +196,7 @@ defmodule Mix.Tasks.Phoenix.New do
                static_deps_prefix: static_deps_prefix,
                brunch: brunch,
                ecto: ecto,
+	       html: html,
```
Comment (josevalim) [inline installer/lib/phoenix_new.ex]: indententation

File: installer/templates/new/lib/application_name/endpoint.ex [hunk 3]
```diff
@@ -1,7 +1,7 @@
 defmodule <%= application_module %>.Endpoint do
   use Phoenix.Endpoint, otp_app: :<%= application_name %>
 
-  socket "/socket", <%= application_module %>.UserSocket
+  socket "/socket", <%= application_module %>.UserSocket<%= if html do %>
```
Comment (josevalim) [inline installer/templates/new/lib/application_name/endpoint.ex]: I would keep Plug.Static regardless. :)

File: installer/templates/new/lib/application_name/endpoint.ex [hunk 4]
```diff
@@ -9,13 +9,13 @@ defmodule <%= application_module %>.Endpoint do
   # when deploying your static files in production.
   plug Plug.Static,
     at: "/", from: :<%= application_name %>, gzip: false,
-    only: ~w(css fonts images js favicon.ico robots.txt)
+    only: ~w(css fonts images js favicon.ico robots.txt)<% end %>
 
   # Code reloading can be explicitly enabled under the
   # :code_reloader configuration of your endpoint.
-  if code_reloading? do
+  if code_reloading? do<%= if html do %>
```
Comment (josevalim) [inline installer/templates/new/lib/application_name/endpoint.ex]: Sweet.

File: installer/templates/new/mix.exs [hunk 5]
```diff
@@ -33,9 +33,9 @@ defmodule <%= application_module %>.Mixfile do
   defp deps do
     [<%= phoenix_dep %>,<%= if ecto do %>
      {:phoenix_ecto, "~> 1.1"},
-     {<%= inspect adapter_app %>, ">= 0.0.0"},<% end %>
+     {<%= inspect adapter_app %>, ">= 0.0.0"},<% end %><%= if html do %>
```
Comment (josevalim) [inline installer/templates/new/mix.exs]: Double sweet.

**Comment by josevalim** [comment]
Thank you @stevedomin! I have added some comments. I think you forgot a "Phoenix.HTML" inside `web/web.ex`. You can test your new changes inside the installer directory:

    cd installer
    mix phoenix.new demo --dev --no-html
    mix test

If the tests pass, we are good to go. :)

File: installer/lib/phoenix_new.ex (hunk 1 above)
Comment (stevedomin) [inline installer/lib/phoenix_new.ex]: That makes sense.

Part of me like it, another part feels like we already have a lot of flags.

Comment (josevalim) [inline installer/lib/phoenix_new.ex]: Let's keep the `--api` flag out for now, we can start another discussion for it. :)

**Comment by stevedomin** [comment]
@josevalim this what the view function look like in `web/web.ex` with the --no-html flag.

```
def view do
    quote do
      use Phoenix.View, root: "web/templates"

      # Import convenience functions from controllers
      import Phoenix.Controller, only: [get_csrf_token: 0, get_flash: 2, view_module: 1]

      import PhotoBlog.Router.Helpers
    end
  end
```

Do you think we should remove that as well?

```
use Phoenix.View, root: "web/templates"

# Import convenience functions from controllers
import Phoenix.Controller, only: [get_csrf_token: 0, get_flash: 2, view_module: 1]
```

**Comment by josevalim** [comment]
No, let's keep them. get_flash is likely not necessary BUT the other two may still be useful, so better to not change because of one import.

**Comment by stevedomin** [comment]
Ok, then I think it's ready :)

File: installer/lib/phoenix_new.ex [hunk 6]
```diff
@@ -57,6 +51,19 @@ defmodule Mix.Tasks.Phoenix.New do
     {:text, "static/robots.txt",              "web/static/assets/robots.txt"},
   ]
 
+  @html [
+    {:eex,  "new/test/controllers/page_controller_test.exs", "test/controllers/page_controller_test.exs"},
+    {:eex,  "new/test/views/error_view_test.exs",            "test/views/error_view_test.exs"},
```
Comment (josevalim) [inline installer/lib/phoenix_new.ex]: We also use the error view for json. I would just generate it as is.

File: installer/lib/phoenix_new.ex [hunk 7]
```diff
@@ -57,6 +51,19 @@ defmodule Mix.Tasks.Phoenix.New do
...
+    {:eex,  "new/test/views/layout_view_test.exs",           "test/views/layout_view_test.exs"},
+    {:eex,  "new/test/views/page_view_test.exs",             "test/views/page_view_test.exs"},
+    {:eex,  "new/web/controllers/page_controller.ex",        "web/controllers/page_controller.ex"},
+    {:eex,  "new/web/templates/layout/app.html.eex",         "web/templates/layout/app.html.eex"},
+    {:eex,  "new/web/templates/page/index.html.eex",         "web/templates/page/index.html.eex"},
+    {:eex,  "new/web/views/error_view.ex",                   "web/views/error_view.ex"},
```
Comment (josevalim) [inline installer/lib/phoenix_new.ex]: So we should also keep the error view tests.

**Comment by josevalim** [comment]
I added a couple more comments. We also need to change the config files.

1. this should be only json:

https://github.com/phoenixframework/phoenix/blob/master/installer/templates/new/config/config.exs#L17

2. this can be skipped:

https://github.com/phoenixframework/phoenix/blob/master/installer/templates/new/config/dev.exs#L18-L25

And I think that's all!

**Comment by stevedomin** [comment]
@josevalim I fixed the issues you reported

Not too happy with the error view as it will be trying to render "4|5xx.json" but as it won't find the template it's going to use 500.html. I feel like the installer could use a bit of refactoring, what do you think? I wrote down a few suggestions, I can send to phoenix-core if that's useful.

**Comment by josevalim** [comment]
@stevedomin should we fix the views for json then? generate 404.json and 500.json returning maps?

> I wrote down a few suggestions, I can send to phoenix-core if that's useful.

Please do!

**Comment by stevedomin** [comment]
@josevalim here are the files I changed:

[error_view_test.exs](https://github.com/stevedomin/phoenix/blob/no-html/installer/templates/new/test/views/error_view_test.exs)
[error_view.ex](https://github.com/stevedomin/phoenix/blob/no-html/installer/templates/new/web/views/error_view.ex)
[phoenix_new_test](https://github.com/stevedomin/phoenix/blob/no-html/installer/test/phoenix_new_test.exs)

I'm not sure if that's the correct way of writing the test.

File: installer/templates/new/test/views/error_view_test.exs [hunk 8]
```diff
@@ -17,5 +17,18 @@ defmodule <%= application_module %>.ErrorViewTest do
   test "render any other" do
     assert render_to_string(<%= application_module %>.ErrorView, "505.html", []) ==
            "Server internal error"
+  end<% else %>test "renders 404.json" do
+    assert Poison.decode!(render_to_string(<%= application_module %>.ErrorView, "404.json", [])) ==
```
Comment (josevalim) [inline installer/templates/new/test/views/error_view_test.exs]: In those cases, I would just call `render`, and not `render_to_string`, and compare the returned map with a map. :)

**Comment by josevalim** [comment]
@stevedomin for JS view tests, we can call `render` instead of `render_to_string` and just directly compare maps. :)

**Comment by stevedomin** [comment]
@josevalim of course, that makes total sense... This is now fixed.

**Comment by josevalim** [comment]
@chrismccord this is ready to ship. Please review and merge :)

**Comment by josevalim** [comment]
@stevedomin  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by optikfluffel** [comment]
You're awesome

File: installer/templates/new/web/views/error_view.ex [hunk 9]
```diff
@@ -13,5 +13,17 @@ defmodule <%= application_module %>.ErrorView do
   # template is found, let's render it as 500
   def template_not_found(_template, assigns) do
     render "500.html", assigns
+  end<% else %>def render("404.json", _assigns) do
+    %{"errors" => %{"detail" => "Page not found"}}
   end
+
+  def render("500.json", _assigns) do
+    %{"errors" => %{"detail" => "Server internal error"}}
```
Comment (chrismccord) [inline installer/templates/new/web/views/error_view.ex]: should we prefer string keys here? My personal preference is atom keys when constructing json maps by hand. Strings everywhere is more consistent, but a bit cumbersome.

**Comment by chrismccord** [comment]
I have one small comment. Once addressed (or not if we're happy with it), this is ready to go!

File: installer/templates/new/web/views/error_view.ex (hunk 9 above)
Comment (josevalim) [inline installer/templates/new/web/views/error_view.ex]: Don't we recommend string keys for payloads in the channel layer? I would return whatever the channel layer returns.

Comment (chrismccord) [inline installer/templates/new/web/views/error_view.ex]: > Don't we recommend string keys for payloads in the channel layer? I would return whatever the channel layer returns.

Not in our doc examples. My general rule is string keys coming in from the outside, and atoms going out

Comment (josevalim) [inline installer/templates/new/web/views/error_view.ex]: Ok! @stevedomin sorry for one extra back and forth but can you please change them to use atom keys and fix the generated tests accordingly? Thank you!

**Comment by fbjork** [comment]
Can't wait to see this work merged. Nice work @stevedomin!

**Comment by chrismccord** [comment]
<3 <3 <3

**Comment by stevedomin** [comment]
Thank you for merging @chrismccord, was on holidays so couldn't change the keys to atoms :)

**Comment by emilsoman** [comment]
@stevedomin I generated an app with `--no-html` , now when I tried to run the app with `mix phoenix.server`, I get the following error:

```
== Compilation error on file web/views/error_view.ex ==
** (CompileError) web/views/error_view.ex:2: module MyApp.ErrorHelpers is not loaded and could not be found
    expanding macro: MyApp.Web.__using__/1
    web/views/error_view.ex:2: MyApp.ErrorView (module)
    (elixir) expanding macro: Kernel.use/2
    web/views/error_view.ex:2: MyApp.ErrorView (module)
    (elixir) lib/kernel/parallel_compiler.ex:100: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/8
```

If I remove the `web/views` directory, it works fine. I'm using phoenix v1.1.0. Here's the full command that I used to generate the app:

```
mix phoenix.new my_app --no-ecto --no-brunch --no-html --module MyApp
```

Here's a link to the repo containing the source: https://github.com/emilsoman/my_app . Am I doing anything wrong ?

**Pull request merged**